### PR TITLE
Add support for code-managed content-types

### PR DESCRIPTION
**Pull request by CrestApps**
Accommodate #11736

Adds `DisabledUserInterface` setting to `ContentTypeSettings` and `ContentPartSettings` which will be used by the UI to prevent one from managing it.

Here is an example on how these settings can be set

```
public class TestMigrations : DataMigration
{
    private readonly IContentDefinitionManager _contentDefinitionManager;

    public TestMigrations(IContentDefinitionManager contentDefinitionManager)
    {
        _contentDefinitionManager = contentDefinitionManager;
    }

    public int Create()
    {
        _contentDefinitionManager.AlterPartDefinition("CodeGeneratedPart", part => part
            .Attachable()
            .DisabledUserInterface() // This is a new extendion
            .WithDisplayName("CodeGeneratedPart")
            .WithDescription("Should not show up on the UI to be managed. It should be attachedable for available type.")
            .WithField("Message", field => field
                .OfType("TextField")
                .WithDisplayName("Message")
            )
        );

        _contentDefinitionManager.AlterTypeDefinition("CodeGeneratedType", type => type
            .Listable()
            .Creatable()
            .DisabledUserInterface() // This is a new extendion
            .DisplayedAs("Code Generated Type")
            .WithPart("TitlePart")
            .WithPart("CodeGeneratedPart")
        );

        return 1;
    }
}
```

I am up for a better name for `DisabledUserInterface()`, let me know if there is a better suggestion? We could change it to `CodeManaged()` or something else

### CODE REVIEW

**Comment by ns8482e** [comment]
@CrestApps Don't like this approach to a settings property. 

`ContentDefinitions` are managed in UI and via migrations.  
IMHO If you really need to have code managed content types then it should not be sourced from `ContentDefinition.json`

**Comment by CrestApps** [comment]
@ns8482e what is the reason you don't like to source it in `ContentDefinition`, it is a definition just like anything else. You'll be able to manage it just like anything else using migration but restrict the UI?

**Comment by ns8482e** [comment]
It's different by nature - `ContentDefinition.json` is mutated at runtime whereas code based parts/items are immutable once it's built hence it doesn't add any value benefit saving it to `ContentDefinition.json`, so does the settings that hides ui

**Comment by CrestApps** [comment]
The idea is to be able to prevent a specific content-definitions from being altered using the UI. If we don't store the settings in the content-definition, where else would you suggest to store it? 

If a module need to alter the content-definition using code, then its free to do "by design it can." There is no restriction to code.

File: src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Metadata/Settings/ContentTypeSettings.cs [hunk 1]
```diff
@@ -31,5 +31,11 @@ public class ContentTypeSettings
         /// Used to determine if this content type supports custom permissions
         /// </summary>
         public bool Securable { get; set; }
+
+
+        /// <summary>
+        /// Used to determine if this content type is managable from the User interface.
```
Comment (agriffard) [inline src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Metadata/Settings/ContentTypeSettings.cs:37]: managable => manageable

File: src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Metadata/Settings/ContentPartSettings.cs [hunk 2]
```diff
@@ -26,5 +26,10 @@ public class ContentPartSettings
         /// Gets or sets the default position of the part when attached to a type.
         /// </summary>
         public string DefaultPosition { get; set; }
+
+        /// <summary>
+        /// Gets or sets whether this part can is managable from the User interface.
```
[the remark this answers is not in the document]
Comment (CrestApps) [reply src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Metadata/Settings/ContentPartSettings.cs]: removed in favor of using permission as per the community recommendation

File: src/OrchardCore.Modules/OrchardCore.ContentTypes/Permissions.cs [hunk 3]
```diff
@@ -26,11 +40,34 @@ public IEnumerable<PermissionStereotype> GetDefaultStereotypes()
...
-            return new[]
+            var permissions = new[]
             {
                 ViewContentTypes,
                 EditContentTypes
             };
+
+            return permissions
+                    .Union(_contentDefinitionManager.ListTypeDefinitions().Select(type => CreatePermissionForType(type)))
+                    .Union(_contentDefinitionManager.LoadPartDefinitions().Select(part => CreatePermissionForPart(part)))
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.ContentTypes/Permissions.cs:51]:
```suggestion
                    .Union(_contentDefinitionManager.ListPartDefinitions().Select(part => CreatePermissionForPart(part)))
```

File: src/OrchardCore.Modules/OrchardCore.ContentTypes/Controllers/AdminController.cs [hunk 4]
```diff
@@ -143,14 +142,14 @@ public async Task<ActionResult> CreatePOST(CreateTypeViewModel viewModel)
 
             await _notifier.SuccessAsync(H["The \"{0}\" content type has been created.", typeViewModel.DisplayName]);
 
-            return RedirectToAction("AddPartsTo", new { id = typeViewModel.Name });
+            return RedirectToAction(nameof(AddPartsTo), new { id = typeViewModel.Name });
         }
 
         public async Task<ActionResult> Edit(string id)
         {
-            if (!await _authorizationService.AuthorizeAsync(User, Permissions.EditContentTypes))
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.ContentTypes/Controllers/AdminController.cs:151]: Keep it, and handle the custom permission in the authorization handler.

File: src/OrchardCore.Modules/OrchardCore.ContentTypes/Permissions.cs [hunk 5]
```diff
@@ -1,13 +1,27 @@
...
 {
     public class Permissions : IPermissionProvider
     {
-        public static readonly Permission ViewContentTypes = new Permission("ViewContentTypes", "View content types.");
-        public static readonly Permission EditContentTypes = new Permission("EditContentTypes", "Edit content types.", isSecurityCritical: true);
+        public static readonly Permission EditContentTypes = new Permission("EditContentTypes", "Edit all content types.", isSecurityCritical: true);
+        public static readonly Permission ViewContentTypes = new Permission("ViewContentTypes", "View content types.", new[] { EditContentTypes });
+
+        private static readonly Permission EditContentType = new Permission("EditContentType_{0}", "Edit content type - {0}", new[] { EditContentTypes });
+        private static readonly Permission EditContentPart = new Permission("EditContentPart_{0}", "Edit content part - {0}", new[] { EditContentTypes });
```
[the remark this answers is not in the document]
Comment (CrestApps) [reply src/OrchardCore.Modules/OrchardCore.ContentTypes/Permissions.cs]: done.

File: src/OrchardCore.Modules/OrchardCore.ContentTypes/Permissions.cs (hunk 3 above)
Comment (CrestApps) [reply src/OrchardCore.Modules/OrchardCore.ContentTypes/Permissions.cs]: done

File: src/OrchardCore.Modules/OrchardCore.ContentTypes/Controllers/AdminController.cs (hunk 4 above)
Comment (CrestApps) [reply src/OrchardCore.Modules/OrchardCore.ContentTypes/Controllers/AdminController.cs:151]: Created a new ContentDefinitionAutherizationHandler to validate content types/parts.

**Comment by CrestApps** [comment]
@sebastienros i don’t think this is going to be useful for anyone. I don’t see a real way to restricting an admin from making a change to a content type since the change can come from a recipe not just UI. I am going to close this PR. If you see a value to this PR, feel free to open or if changes are needed I’ll still support it.

**Pull request closed without merging**